### PR TITLE
advent-2016: add 'context' post

### PR DESCRIPTION
**Pull request by samv**
This post was originally posted in multiple parts and requests for
comments drawn from a couple of core authorities on the topic.  This
is a revision based on comments and feedback primarily from Peter
Bourgon.

For example, the feedback that storing 'common log objects' in context
might end up giving users the wrong idea, even though they are
log-scoped, was taken to heart: this code was already trimmed down
from my real code where I did that, so I could simplify the example to
just store the IDs and still serve the purpose of illustration.

Finally, the original plan to include a summary of three applications
of context in this piece with outlinks back to the full medium posts
were instead changed to include the introduction and the first piece,
based on feedback and engagement with the first "introduction" piece I
posted.

### CODE REVIEW

File: content/advent-2016/context-logging.md [hunk 1]
```diff
@@ -0,0 +1,580 @@
...
+
+## That sounds great!  Now how will this make my logs better?
+
+There's many ways to make sure that a service is operating correctly
+under the hood.  Let's talk about one standard approach - logging -
+and how `context` can help you turn your logs into rich sources of
+insight.
+
+Initially, you might start with the built-in `log` library, and its
+`Printf` inteface:
```
Comment (bradleyfalzon) [inline content/advent-2016/context-logging.md]: inteface should be interface?

File: content/advent-2016/context-logging.md [hunk 2]
```diff
@@ -0,0 +1,580 @@
...
+
+When you solved this problem in
+[blub](http://wiki.c2.com/?BlubParadox), it was pretty easy: you just
+assigned a request ID to the request, and then some magic library made
+that request ID log in every line logged anywhere in the program.
+
+So how can we do this in go?
+
+First, let's ditch the standard `log` library, which is horribly
+unstructured, and use Uber's `zap` logger.  You could equally well do
```
Comment (bradleyfalzon) [inline content/advent-2016/context-logging.md]: use should be uses ?

File: content/advent-2016/context-logging.md [hunk 3]
```diff
@@ -0,0 +1,580 @@
...
+
+It doesn't matter an awful lot.  It could be a regular HTTP service of
+some kind, a consumer of kafka topics, a cog in a vast stream
+processing machine, anything - the key characteristic being that a
+single go program is responsible for processing more than one
+independent thing at a time, or at some point you would want to switch
+it from processing one thing at a time to many things at the same
+time.  Why else are you using go?
+
+## Enough future tense.  Give me some content.  What does `context` do ?
```
Comment (joncalhoun) [inline content/advent-2016/context-logging.md]: Probably didn't intend for the space between "do" and "?"

Comment (samv) [inline content/advent-2016/context-logging.md]: Thanks, fixed!

**Comment by samv** [comment]
Thanks for the proof-reading, @bradleyfalzon and @joncalhoun - typos are fixed.

File: content/advent-2016/context-logging.md [hunk 4]
```diff
@@ -0,0 +1,580 @@
...
+            }
+            if ctxSessionId, ok := ctx.Value(sessionIdKey).(string); ok {
+               newLogger = newLogger.With(zap.String("sessionId", ctxSessionId))
+            }
+        }
+        return newLogger
+    }
+```
+
+Now, the program can log with the request ID, just by using this
```
Comment (nathany) [inline content/advent-2016/context-logging.md]: Thanks for the article. Btw, I think this should be `myappcontext.Logger`.

File: content/advent-2016/context-logging.md [hunk 5]
```diff
@@ -0,0 +1,580 @@
...
+    )
+
+    var httpContext = context.Background()
+    
+    func RequestHandler(w http.ResponseWriter, r *http.Request) {
+        rqId := uuid.NewRandom()
+        rqCtx := myappcontext.WithRqId(httpContext, rqId)
+        logger = myappctx.Logger(rqCtx)
+        logger.Info("handling /doit request", zap.String("for", user))
+        somelbrary.DoSomethingAwesome(rqCtx, ...)
```
Comment (nathany) [inline content/advent-2016/context-logging.md]: typo: somelibrary

File: content/advent-2016/context-logging.md [hunk 6]
```diff
@@ -0,0 +1,580 @@
...
+
+Values not being set in the context is explicitly encouraged. If you
+try to fetch from the context and the key is not set, you'll get `nil`
+back instead.
+
+Did I say "encouraged"?  What I actually mean is, if your function
+doesn't work at all unless magic values are poked into the context
+first, you are almost certainly abusing `context`.
+
+You should always try to cast the value back to the type you expected
```
Comment (nathany) [inline content/advent-2016/context-logging.md]: technically this is called a "type assertion", not "casting"
https://golang.org/ref/spec#Type_assertions

File: content/advent-2016/context-logging.md [hunk 7]
```diff
@@ -0,0 +1,580 @@
...
+
+And then, you'll finally have matched Blub for its logging ability.
+But at 1000 times the speed!  :)
+
+## Anti-Patterns
+
+As `context` is still relatively new, and powerful, people are using
+it in ways not intended by the original authors.  Not that this sort
+of thing is always bad, but here are a few of the common abuses of the
+module, along with a brief argument about why you shouldn't do this.
```
Comment (nathany) [inline content/advent-2016/context-logging.md]: ~~module~~ package

File: content/advent-2016/context-logging.md [hunk 8]
```diff
@@ -0,0 +1,580 @@
...
+higher level.  Resist the temptation to just throw it in the context:
+this hides what should be an explicit, required parameter.
+
+**Why this is bad**: readers of your code should be able to tell by
+looking at the function arguments key factors determining the
+operation of the function.  By passing required arguments via
+`context`, you're writing code which is surprising to first-time
+readers of it.
+
+**What you should do instead**: you'll have to pass that function down
```
Comment (nathany) [inline content/advent-2016/context-logging.md]: ~~function~~ down
argument down

File: content/advent-2016/context-logging.md [hunk 9]
```diff
@@ -0,0 +1,580 @@
...
+which can just grab all the extra arguments you passed on a function
+call.  It commonly gets used to pass functions down several scopes, as
+functions pull the parts out they want as they go.
+
+**Why this is bad**: In python, the string keys in Python mean that
+the arguments can potentially collide.  Every function has to handle
+all the arguments in between, and because the keys are all strings you
+don't know who is going to do what with that extra argument.  In Go,
+using key types which are private avoids that problem, but again this
+comes down to - this is surprising to the reader.  `context` is
```
Comment (nathany) [inline content/advent-2016/context-logging.md]: "surprising to the reader."

who, me? 😜

**Pull request merged**